working/augmentation-libraries/feature-specification.md

```diff
@@ -321,82 +322,150 @@ doesn't, it must repeat the type from the augmented definition__.
 
 ### Augmented Expression
 
-The exact result of an `augmented` expression depends on what is being
-augmented, but it generally follows the same rules as any normal identifier:
-
-*   **Augmenting getters**: Within an augmenting getter `augmented` invokes
-    the augmented getter and evaluates to its return value. If augmenting a
-    variable with a getter, this will invoke the implicitly induced getter
-    from the augmented variable declaration.
-
-*   **Augmenting setters**: Within an augmenting setter `augmented` must be
-    followed by an `=` and will directly invoke the augmented setter. If
-    augmenting a variable with a setter, this will invoke the implicitly
-    induced setter from the augmented variable declaration.
-
-*   **Augmenting fields**: Within an augmenting variable declaration,
-    `augmented` can only be used in an initializer expression, and refers
-    to the augmented variable's initializing expression, which is
-    immediately evaluated.
-
-    It is a compile-time error to use `augmented` in an augmenting
-    variable's initializer if the member being augmented is not a variable
-    declaration with an initializing expression.
+The identifier `augmented` is a reserved word, with associated special syntax,
+when it occurs inside an expression or body of an augmenting variable, getter,
+setter, function, operator or constructor declaration.
+
+The exact grammar and meaning of the `augmented` keyword depends on what is
+being augmented.
+
+*   **Augmenting variables**: Within an augmenting variable declaration's
+    initializer expression, `augmented` can only be used in an initializer
+    expression, and it refers to the augmented variable's initializing
+    expression. The reserved words parses as a `<primary>` expression.
+    Evaluating `augmented` evaluates the augmented declaration's initializer
+    expression, and evaluates to the same result.
+    The static type of `augmented` is the static type of the initializer
+    expression of the augmented declaration.
+
+*   **Augmenting getters**: Within an augmenting getter's body, the `augmented`
+    expression invokes the augmented getter and evaluates to its return value.
+    When augmenting a variable with a getter, this will invoke the implicitly
+    induced getter of the augmented variable declaration.
+    _This applies even to `late` variable declarations, where the implicit
+    getter does more than just access the underlying storage._
+    The reserved words parses as a `<primary>` expression.
+    Evaluating `augmented` invokes the augmented declaration's getter
+    function, and evaluates to the returned value.
+    The static type of `augmented` is the declared return type of the augmented
```

Comment:
Should we specify this as the declared return type or something a bit more general, like how we use the "static type" for initializer expressions?

For instance, does "declared return type" cover the override case correctly?

Reply:
Good point. We want the return type, not the declared return type. Or maybe we want the inferred return type. If there is a difference, it's the type that the declaration is determined to have after (top-level?) type inference has been performed.

We just need to define what the return type of a getter declaration is.

The getter return type of a non-augmenting getter declaration is the type denoted by its declared return type, if any.
If there is no declared return type, the getter is an instance getter of a `class`, `mixin` or `enum` declaration, and any superinterface of the surrounding type has a getter with the same name, then the return type is the return type of the getter with that name in the combined superinterface of the current declaration. _It's a compile-time error if the combined superinterface does not have a combined type signature for that getter._
Otherwise the return type of the getter is `dynamic`. 
(We don't infer the type of bodies in top-level inference, right? It's only for variables.)

The getter return type of a variable declaration is the type denoted by the declared type of the variable, if any. Otherwise, if the variable has no initializer and it's an instance variable, it's the return type of the getter or parameter type of the setter in the combined super-interface, if any. Otherwise it's the static type of its initializer expression.

The getter return type of an augmenting getter declaration is the getter return type of the augmented declaration. (Can't change the type. If the augmenting declaration writes a type, it must be that type.)

It gets complicated, because it's basically repeating all of type inference. I hope there is a way to reuse type inference instead.

Reply:
Right, but I think all we need to state here is that it uses the inferred return type? We don't need to explain how the inference works.